File: iot_driver_copilot/decoder_(matrix_decoder_series)/driver.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <map>
#include <thread>
#include <vector>
#include <atomic>
#include <mutex>
#include <csignal>
#include <cstdlib>
#include <ctime>
#include <chrono>
#include <functional>
#include <condition_variable>

// External Libraries
#include <yaml-cpp/yaml.h>
#include <nlohmann/json.hpp>
#include <paho.mqtt.cpp/async_client.h>

// Kubernetes
#include <kubernetes/client/client.h> // Placeholder: use official or supported k8s C++ client

// HTTP server: use Pistache for simplicity
#include <pistache/endpoint.h>
#include <pistache/router.h>
#include <pistache/http.h>

// Logging
#include <spdlog/spdlog.h>
#include <spdlog/sinks/stdout_color_sinks.h>

using json = nlohmann::json;
using namespace std::chrono_literals;

using std::string;
using std::map;
using std::vector;
using std::thread;
using std::mutex;
using std::lock_guard;
using std::unique_lock;
using std::condition_variable;

using std::atomic_bool;
using std::atomic_int;

//////////////////////////////////////////
// Utility Functions
//////////////////////////////////////////

inline string getenv_or_default(const string &var, const string &def) {
    const char *val = std::getenv(var.c_str());
    return val ? string(val) : def;
}

inline int getenv_or_default_int(const string &var, int def) {
    const char *val = std::getenv(var.c_str());
    return val ? std::stoi(val) : def;
}

//////////////////////////////////////////
// ShifuClient: Handles K8s interaction
//////////////////////////////////////////

class ShifuClient {
public:
    ShifuClient(
        const string &device_name,
        const string &namespace_,
        const string &config_mount_path,
        std::shared_ptr<spdlog::logger> logger)
        : device_name_(device_name),
          namespace_(namespace_),
          config_mount_path_(config_mount_path),
          logger_(logger) {
        _init_k8s_client();
    }

    void _init_k8s_client() {
        try {
            // Initialize k8s client, in-cluster or local kubeconfig
            // Placeholder: Use kubernetes::client::Client from cpp-k8s-client
            // If in cluster: use default, else fall back to KUBECONFIG
            logger_->info("Initializing Kubernetes client...");
            // k8s_client_ = std::make_shared<kubernetes::client::Client>();
            // Actual implementation depends on library
        } catch (const std::exception &e) {
            logger_->error("Failed to initialize Kubernetes client: {}", e.what());
            // Continue as read-only driver if no client
        }
    }

    json get_edge_device() {
        // Placeholder: fetch EdgeDevice CR
        try {
            // auto resource = k8s_client_->get_edgedevice(device_name_, namespace_);
            // return resource;
            return json(); // Return empty for now
        } catch (const std::exception &e) {
            logger_->warn("get_edge_device: {}", e.what());
            return json();
        }
    }

    string get_device_address() {
        // Placeholder: extract address from EdgeDevice CR status
        try {
            json edge_device = get_edge_device();
            if (!edge_device.is_null() && edge_device.contains("status") && edge_device["status"].contains("address")) {
                return edge_device["status"]["address"];
            }
        } catch (const std::exception &e) {
            logger_->warn("get_device_address: {}", e.what());
        }
        return "";
    }

    void update_device_status(const string &status) {
        // Placeholder: PATCH or UPDATE EdgeDevice status subresource
        try {
            logger_->info("Updating device status in K8s: {}", status);
            // k8s_client_->update_edgedevice_status(device_name_, namespace_, status);
        } catch (const std::exception &e) {
            logger_->warn("update_device_status: {}", e.what());
        }
    }

    string read_mounted_config_file(const string &filename) {
        string path = config_mount_path_ + "/" + filename;
        std::ifstream file(path);
        if (!file) {
            logger_->warn("Cannot open config file: {}", path);
            return "";
        }
        std::stringstream buffer;
        buffer << file.rdbuf();
        return buffer.str();
    }

    YAML::Node get_instruction_config() {
        string content = read_mounted_config_file("instructions");
        if (content.empty()) {
            logger_->warn("Instruction config is empty.");
            return YAML::Node();
        }
        try {
            return YAML::Load(content);
        } catch (const std::exception &e) {
            logger_->error("YAML parsing error: {}", e.what());
            return YAML::Node();
        }
    }

private:
    string device_name_;
    string namespace_;
    string config_mount_path_;
    std::shared_ptr<spdlog::logger> logger_;
    // std::shared_ptr<kubernetes::client::Client> k8s_client_;
};

//////////////////////////////////////////
// Device SDK Abstraction (Stub)
//////////////////////////////////////////

class DecoderSDK {
public:
    explicit DecoderSDK(std::shared_ptr<spdlog::logger> logger)
        : connected_(false), logger_(logger) {}

    bool connect(const string &ip, int port, const string &username, const string &password) {
        // Simulate device login
        logger_->info("Connecting to decoder device at {}:{}...", ip, port);
        std::this_thread::sleep_for(500ms);
        connected_ = true;
        logger_->info("Decoder device connected.");
        return true;
    }
    void disconnect() {
        logger_->info("Disconnecting from decoder device...");
        connected_ = false;
    }
    bool is_connected() const { return connected_; }

    // Simulated SDK commands (replace with actual SDK calls)
    json get_status() {
        // Return dummy device status
        json j = {
            {"deviceStatus", "online"},
            {"channels", {
                {"channel1", {{"status", "active"}}},
                {"channel2", {{"status", "inactive"}}}
            }},
            {"alarms", {{"code", 0}, {"description", "none"}}}
        };
        return j;
    }

    json decode_control(const json &payload) {
        // Simulate decode start/stop
        string action = payload.value("action", "");
        logger_->info("DecoderSDK decode_control action: {}", action);
        if (action == "start") {
            return {{"result", "started"}};
        } else if (action == "stop") {
            return {{"result", "stopped"}};
        }
        return {{"result", "unknown_action"}};
    }

    json config_update(const json &payload) {
        logger_->info("DecoderSDK config_update: {}", payload.dump());
        return {{"result", "ok"}};
    }

    json login(const json &payload) {
        logger_->info("DecoderSDK login: {}", payload.dump());
        return {{"result", "login_success"}};
    }

    json logout(const json &payload) {
        logger_->info("DecoderSDK logout: {}", payload.dump());
        return {{"result", "logout_success"}};
    }

private:
    bool connected_;
    std::shared_ptr<spdlog::logger> logger_;
};

//////////////////////////////////////////
// Main DeviceShifu Driver
//////////////////////////////////////////

class DeviceShifuDriver {
public:
    DeviceShifuDriver()
      : shutdown_flag_(false),
        logger_(spdlog::stdout_color_mt("console")) {
        configure();
        shifu_client_ = std::make_unique<ShifuClient>(
            device_name_, edgedevice_namespace_, config_mount_path_, logger_);
        decoder_sdk_ = std::make_unique<DecoderSDK>(logger_);
        mqtt_client_ = nullptr;
    }

    void configure() {
        // Set up configuration from environment
        device_name_ = getenv_or_default("EDGEDEVICE_NAME", "deviceshifu-decoder");
        edgedevice_namespace_ = getenv_or_default("EDGEDEVICE_NAMESPACE", "devices");
        config_mount_path_ = getenv_or_default("CONFIG_MOUNT_PATH", "/etc/edgedevice/config");
        mqtt_broker_ = getenv_or_default("MQTT_BROKER", "127.0.0.1");
        mqtt_port_ = getenv_or_default_int("MQTT_BROKER_PORT", 1883);
        mqtt_username_ = getenv_or_default("MQTT_BROKER_USERNAME", "");
        mqtt_password_ = getenv_or_default("MQTT_BROKER_PASSWORD", "");
        mqtt_topic_prefix_ = getenv_or_default("MQTT_TOPIC_PREFIX", "shifu");
        http_host_ = getenv_or_default("HTTP_HOST", "0.0.0.0");
        http_port_ = getenv_or_default_int("HTTP_PORT", 8080);
        log_level_ = getenv_or_default("LOG_LEVEL", "info");
        if (log_level_ == "debug") {
            logger_->set_level(spdlog::level::debug);
        } else if (log_level_ == "info") {
            logger_->set_level(spdlog::level::info);
        } else if (log_level_ == "warn") {
            logger_->set_level(spdlog::level::warn);
        } else {
            logger_->set_level(spdlog::level::info);
        }
    }

    void run() {
        logger_->info("Starting DeviceShifu MQTT Driver for '{}'", device_name_);
        auto config_node = shifu_client_->get_instruction_config();
        if (!config_node) {
            logger_->error("Instruction config not found. Exiting.");
            return;
        }

        parse_instructions(config_node);

        if (!connect_device()) {
            logger_->error("Failed to connect to decoder device. Exiting.");
            return;
        }

        connect_mqtt();

        start_http_server();

        start_periodic_publishers();

        // Signal handling for graceful shutdown
        std::signal(SIGINT, signal_handler_static);
        std::signal(SIGTERM, signal_handler_static);

        // Main event loop
        while (!shutdown_flag_) {
            std::this_thread::sleep_for(1s);
        }
        shutdown();
    }

    static void signal_handler_static(int signum) {
        instance_->signal_handler(signum);
    }

    void signal_handler(int signum) {
        logger_->warn("Received signal {}, shutting down...", signum);
        shutdown_flag_ = true;
        cv_shutdown_.notify_all();
    }

    void shutdown() {
        logger_->info("Shutting down DeviceShifu MQTT driver...");
        if (mqtt_client_) {
            try {
                mqtt_client_->disconnect()->wait();
            } catch (...) {
                logger_->warn("MQTT disconnect error");
            }
        }
        if (http_endpoint_) {
            http_endpoint_->shutdown();
        }
        if (decoder_sdk_) {
            decoder_sdk_->disconnect();
        }
        for (auto &t : publisher_threads_) {
            if (t.joinable())
                t.join();
        }
        logger_->info("Shutdown complete.");
    }

private:
    // Configuration
    string device_name_;
    string edgedevice_namespace_;
    string config_mount_path_;
    string mqtt_broker_;
    int mqtt_port_;
    string mqtt_username_;
    string mqtt_password_;
    string mqtt_topic_prefix_;
    string http_host_;
    int http_port_;
    string log_level_;

    // State
    std::unique_ptr<ShifuClient> shifu_client_;
    std::unique_ptr<DecoderSDK> decoder_sdk_;
    std::shared_ptr<spdlog::logger> logger_;

    // MQTT
    std::unique_ptr<mqtt::async_client> mqtt_client_;
    mqtt::connect_options mqtt_conn_opts_;
    std::atomic_bool mqtt_connected_{false};
    std::mutex mqtt_mutex_;

    // Publishers
    struct InstructionInfo {
        string name;
        string method;
        string path;
        string mode;
        int publish_interval_ms;
        int qos;
    };
    map<string, InstructionInfo> instructions_;
    map<string, json> latest_data_;
    std::mutex data_mutex_;

    vector<thread> publisher_threads_;
    atomic_bool shutdown_flag_;
    condition_variable cv_shutdown_;

    // HTTP
    std::shared_ptr<Pistache::Http::Endpoint> http_endpoint_;

    // Singleton instance for signal handling
    static DeviceShifuDriver *instance_;

    //////////////////////////////////
    // Device/Instruction Handling
    //////////////////////////////////

    bool connect_device() {
        // Fetch device address and credentials (simulate)
        string address = shifu_client_->get_device_address();
        if (address.empty())
            address = "127.0.0.1";
        int port = 8000; // Simulate
        string username = "admin";
        string password = "admin123";
        return decoder_sdk_->connect(address, port, username, password);
    }

    void parse_instructions(const YAML::Node &config_node) {
        try {
            for (const auto &inst : config_node) {
                string name = inst.first.as<string>();
                const auto &node = inst.second;
                InstructionInfo info;
                info.name = name;
                info.method = node["method"] ? node["method"].as<string>() : "PUBLISH";
                info.path = node["path"] ? node["path"].as<string>() : "";
                info.mode = node["mode"] ? node["mode"].as<string>() : "publisher";
                info.publish_interval_ms = node["publishIntervalMS"] ? node["publishIntervalMS"].as<int>() : 0;
                info.qos = node["qos"] ? node["qos"].as<int>() : 0;
                instructions_[name] = info;
            }
        } catch (const std::exception &e) {
            logger_->error("parse_instructions: {}", e.what());
        }
    }

    void start_periodic_publishers() {
        for (const auto &pair : instructions_) {
            const auto &info = pair.second;
            if (info.mode == "publisher" && info.publish_interval_ms > 0) {
                publisher_threads_.emplace_back(&DeviceShifuDriver::publish_topic_periodically, this, info);
            }
        }
    }

    void publish_topic_periodically(InstructionInfo info) {
        logger_->info("Starting periodic publisher for '{}' every {} ms", info.name, info.publish_interval_ms);
        while (!shutdown_flag_) {
            try {
                publish_instruction(info);
            } catch (const std::exception &e) {
                logger_->warn("Publish thread error [{}]: {}", info.name, e.what());
            }
            std::unique_lock<std::mutex> lk(data_mutex_);
            cv_shutdown_.wait_for(lk, std::chrono::milliseconds(info.publish_interval_ms), [this]() { return shutdown_flag_; });
        }
        logger_->info("Publisher thread for '{}' exiting", info.name);
    }

    void publish_instruction(const InstructionInfo &info) {
        json payload;
        if (info.name == "status") {
            payload = decoder_sdk_->get_status();
        }
        // Add more mappings as needed.

        {
            lock_guard<mutex> lock(data_mutex_);
            latest_data_[info.name] = payload;
        }
        string topic = mqtt_topic_prefix_ + "/" + device_name_ + "/" + info.name;
        publish_mqtt(topic, payload.dump(), info.qos);
    }

    //////////////////////////////////
    // MQTT Handling
    //////////////////////////////////

    void connect_mqtt() {
        string client_id = device_name_ + "-" + std::to_string(std::time(nullptr));
        string server_uri = "tcp://" + mqtt_broker_ + ":" + std::to_string(mqtt_port_);

        mqtt_client_ = std::make_unique<mqtt::async_client>(server_uri, client_id);

        mqtt_conn_opts_.set_keep_alive_interval(20);
        mqtt_conn_opts_.set_clean_session(true);
        if (!mqtt_username_.empty())
            mqtt_conn_opts_.set_user_name(mqtt_username_);
        if (!mqtt_password_.empty())
            mqtt_conn_opts_.set_password(mqtt_password_);

        mqtt_client_->set_connected_handler([this](const string &) {
            mqtt_connected_ = true;
            logger_->info("MQTT connected");
            subscribe_control_topics();
        });
        mqtt_client_->set_connection_lost_handler([this](const string &) {
            mqtt_connected_ = false;
            logger_->warn("MQTT connection lost");
            reconnect_mqtt();
        });
        mqtt_client_->set_message_callback([this](mqtt::const_message_ptr msg) {
            handle_mqtt_message(msg);
        });

        try {
            mqtt_client_->connect(mqtt_conn_opts_)->wait();
        } catch (const mqtt::exception &e) {
            logger_->error("MQTT connect failed: {}", e.what());
        }
    }

    void reconnect_mqtt() {
        while (!shutdown_flag_ && !mqtt_connected_) {
            logger_->info("Attempting to reconnect to MQTT...");
            try {
                mqtt_client_->reconnect()->wait();
                logger_->info("Reconnected to MQTT.");
                return;
            } catch (...) {
                logger_->warn("MQTT reconnect failed, retrying in 2s...");
                std::this_thread::sleep_for(2s);
            }
        }
    }

    void subscribe_control_topics() {
        // Subscribe to command/control topics
        string control_topic = mqtt_topic_prefix_ + "/" + device_name_ + "/control/#";
        mqtt_client_->subscribe(control_topic, 1)->wait();
        logger_->info("Subscribed to control topic: {}", control_topic);

        // Subscribe to all SUBSCRIBE mode instructions
        for (const auto &pair : instructions_) {
            if (pair.second.method == "SUBSCRIBE") {
                string topic = mqtt_topic_prefix_ + "/" + device_name_ + "/" + pair.second.name;
                mqtt_client_->subscribe(topic, pair.second.qos)->wait();
                logger_->info("Subscribed to instruction topic: {}", topic);
            }
        }
    }

    void handle_mqtt_message(mqtt::const_message_ptr msg) {
        string topic = msg->get_topic();
        string payload = msg->to_string();
        logger_->info("MQTT received: topic='{}' payload='{}'", topic, payload);

        try {
            json j = json::parse(payload);
            // Handle control commands
            if (topic.find("/control/") != string::npos) {
                // Extract command
                string cmd = topic.substr(topic.find("/control/") + 9);
                handle_control_command(cmd, j);
            } else {
                // Handle instruction SUBSCRIBE topics
                for (const auto &pair : instructions_) {
                    string t = mqtt_topic_prefix_ + "/" + device_name_ + "/" + pair.second.name;
                    if (topic == t) {
                        handle_instruction(pair.second, j);
                        break;
                    }
                }
            }
        } catch (const std::exception &e) {
            logger_->warn("MQTT message parse error: {}", e.what());
        }
    }

    void handle_control_command(const string &cmd, const json &payload) {
        logger_->info("Handling control command: {} payload: {}", cmd, payload.dump());
        if (cmd == "login") {
            auto resp = decoder_sdk_->login(payload);
            publish_mqtt(mqtt_topic_prefix_ + "/" + device_name_ + "/status", resp.dump(), 1);
        } else if (cmd == "logout") {
            auto resp = decoder_sdk_->logout(payload);
            publish_mqtt(mqtt_topic_prefix_ + "/" + device_name_ + "/status", resp.dump(), 1);
        } else if (cmd == "decode") {
            auto resp = decoder_sdk_->decode_control(payload);
            publish_mqtt(mqtt_topic_prefix_ + "/" + device_name_ + "/decode/ack", resp.dump(), 1);
        } else if (cmd == "config") {
            auto resp = decoder_sdk_->config_update(payload);
            publish_mqtt(mqtt_topic_prefix_ + "/" + device_name_ + "/config/ack", resp.dump(), 1);
        }
        // Extend with more commands as needed
    }

    void handle_instruction(const InstructionInfo &info, const json &payload) {
        logger_->info("Handling instruction {} with payload {}", info.name, payload.dump());
        // For SUBSCRIBE mode, update latest_data and respond if needed
        {
            lock_guard<mutex> lock(data_mutex_);
            latest_data_[info.name] = payload;
        }
        // Optionally, perform device action here
    }

    void publish_mqtt(const string &topic, const string &payload, int qos) {
        if (!mqtt_connected_)
            return;
        try {
            mqtt::message_ptr msg = mqtt::make_message(topic, payload);
            msg->set_qos(qos);
            mqtt_client_->publish(msg);
            logger_->debug("Published MQTT: topic='{}', payload='{}'", topic, payload);
        } catch (const std::exception &e) {
            logger_->warn("MQTT publish error: {}", e.what());
        }
    }

    //////////////////////////////////
    // HTTP Server (Pistache)
    //////////////////////////////////

    void start_http_server() {
        using namespace Pistache;

        Address addr(http_host_, Port(http_port_));
        http_endpoint_ = std::make_shared<Http::Endpoint>(addr);

        auto opts = Http::Endpoint::options().threads(1);
        http_endpoint_->init(opts);

        Rest::Router router;
        setup_routes(router);

        http_endpoint_->setHandler(router.handler());
        thread([this]() {
            http_endpoint_->serve();
        }).detach();
        logger_->info("HTTP server started at {}:{}", http_host_, http_port_);
    }

    void setup_routes(Pistache::Rest::Router &router) {
        using namespace Pistache::Rest;

        Routes::Get(router, "/health", Routes::bind(&DeviceShifuDriver::route_health, this));
        Routes::Get(router, "/status", Routes::bind(&DeviceShifuDriver::route_status, this));
    }

    void route_health(const Pistache::Rest::Request &, Pistache::Http::ResponseWriter response) {
        response.send(Pistache::Http::Code::Ok, "{\"status\":\"healthy\"}\n", MIME(Application, Json));
    }

    void route_status(const Pistache::Rest::Request &, Pistache::Http::ResponseWriter response) {
        json status;
        {
            lock_guard<mutex> lock(data_mutex_);
            status["device"] = device_name_;
            status["latest_data"] = latest_data_;
            status["mqtt_connected"] = mqtt_connected_;
            status["device_connected"] = decoder_sdk_->is_connected();
        }
        response.send(Pistache::Http::Code::Ok, status.dump(), MIME(Application, Json));
    }
};

DeviceShifuDriver *DeviceShifuDriver::instance_ = nullptr;


//////////////////////////////////////////
// Main Entrypoint
//////////////////////////////////////////

int main() {
    DeviceShifuDriver driver;
    DeviceShifuDriver::instance_ = &driver;
    driver.run();
    return 0;
}

File: iot_driver_copilot/decoder/tv_wall_device/driver.py
import os
import sys
import time
import json
import yaml
import signal
import threading
import logging
from typing import Dict, Any, Optional

import paho.mqtt.client as mqtt
from flask import Flask, jsonify
from kubernetes import client as k8s_client, config as k8s_config
from kubernetes.client.rest import ApiException

# Logging setup
LOG_LEVEL = os.environ.get("LOG_LEVEL", "INFO").upper()
logging.basicConfig(
    level=LOG_LEVEL,
    format="%(asctime)s [%(levelname)s] %(threadName)s %(name)s: %(message)s"
)
logger = logging.getLogger("deviceshifu-mqtt-driver")

# Default env vars
EDGEDEVICE_NAME = os.environ.get("EDGEDEVICE_NAME", "deviceshifu-multichannel-video-decoder")
EDGEDEVICE_NAMESPACE = os.environ.get("EDGEDEVICE_NAMESPACE", "devices")
CONFIG_MOUNT_PATH = os.environ.get("CONFIG_MOUNT_PATH", "/etc/edgedevice/config")
MQTT_BROKER = os.environ.get("MQTT_BROKER", "localhost")
MQTT_BROKER_PORT = int(os.environ.get("MQTT_BROKER_PORT", "1883"))
MQTT_BROKER_USERNAME = os.environ.get("MQTT_BROKER_USERNAME", "")
MQTT_BROKER_PASSWORD = os.environ.get("MQTT_BROKER_PASSWORD", "")
MQTT_TOPIC_PREFIX = os.environ.get("MQTT_TOPIC_PREFIX", "shifu")
HTTP_HOST = os.environ.get("HTTP_HOST", "0.0.0.0")
HTTP_PORT = int(os.environ.get("HTTP_PORT", "8080"))

# ConfigMap filenames
INSTRUCTIONS_FILE = os.path.join(CONFIG_MOUNT_PATH, "instructions")
DRIVER_PROPERTIES_FILE = os.path.join(CONFIG_MOUNT_PATH, "driverProperties")

shutdown_flag = threading.Event()

class ShifuClient:
    def __init__(self, name: str, namespace: str):
        self.name = name
        self.namespace = namespace
        self.k8s_api = self._init_k8s_client()

    def _init_k8s_client(self):
        try:
            k8s_config.load_incluster_config()
            logger.info("Loaded in-cluster Kubernetes config.")
        except Exception:
            try:
                k8s_config.load_kube_config()
                logger.info("Loaded local kubeconfig.")
            except Exception as e:
                logger.error(f"Failed to load Kubernetes config: {e}")
                raise
        return k8s_client.CustomObjectsApi()

    def get_edge_device(self) -> Optional[Dict[str, Any]]:
        group = "shifu.edgenesis.io"
        version = "v1alpha1"
        plural = "edgedevices"
        try:
            ed = self.k8s_api.get_namespaced_custom_object(
                group=group,
                version=version,
                namespace=self.namespace,
                plural=plural,
                name=self.name
            )
            logger.debug(f"EdgeDevice fetched: {ed}")
            return ed
        except ApiException as e:
            logger.error(f"Failed to get EdgeDevice: {e}")
            return None
        except Exception as exc:
            logger.error(f"Unexpected exception getting EdgeDevice: {exc}")
            return None

    def get_device_address(self) -> Optional[str]:
        ed = self.get_edge_device()
        if ed:
            try:
                addr = ed.get("spec", {}).get("address", None)
                logger.info(f"Device address: {addr}")
                return addr
            except Exception as e:
                logger.warning(f"Could not get device address: {e}")
        return None

    def update_device_status(self, status: Dict[str, Any]) -> bool:
        group = "shifu.edgenesis.io"
        version = "v1alpha1"
        plural = "edgedevices"
        body = {"status": status}
        try:
            self.k8s_api.patch_namespaced_custom_object_status(
                group=group,
                version=version,
                namespace=self.namespace,
                plural=plural,
                name=self.name,
                body=body
            )
            logger.info(f"EdgeDevice status updated: {status}")
            return True
        except ApiException as e:
            logger.error(f"Failed to update EdgeDevice status: {e}")
            return False

    def read_mounted_config_file(self, filename: str) -> Optional[Dict[str, Any]]:
        try:
            with open(filename, "r") as f:
                content = yaml.safe_load(f)
                logger.debug(f"Config file '{filename}' loaded: {content}")
                return content
        except Exception as e:
            logger.error(f"Could not read config file {filename}: {e}")
            return None

    def get_instruction_config(self) -> Dict[str, Any]:
        config = self.read_mounted_config_file(INSTRUCTIONS_FILE)
        if not config:
            logger.warning("Instructions config is empty or missing.")
            return {}
        return config

class DeviceShifuMQTTDriver:
    def __init__(self):
        self.device_name = EDGEDEVICE_NAME
        self.device_namespace = EDGEDEVICE_NAMESPACE
        self.shifu_client = ShifuClient(self.device_name, self.device_namespace)
        self.device_address = self.shifu_client.get_device_address()
        self.latest_data: Dict[str, Any] = {}
        self.publish_intervals: Dict[str, int] = {}
        self.publish_threads: Dict[str, threading.Thread] = {}
        self.mqtt_client: Optional[mqtt.Client] = None
        self.connected = False
        self.mqtt_client_id = f"{self.device_name}-{int(time.time())}"
        self.instruction_config = self.shifu_client.get_instruction_config()
        self.driver_properties = self.shifu_client.read_mounted_config_file(DRIVER_PROPERTIES_FILE)
        self._parse_publish_intervals()
        self._flask_app = Flask(__name__)
        self._http_thread = threading.Thread(target=self._run_http_server, name="HTTPServer", daemon=True)
        self._mqtt_thread = threading.Thread(target=self._run_mqtt_loop, name="MQTTLoop", daemon=True)

    def _parse_publish_intervals(self):
        if not self.instruction_config:
            logger.warning("No instruction config to parse publish intervals.")
            return
        for instruction, cfg in self.instruction_config.items():
            if isinstance(cfg, dict) and cfg.get("mode", "publisher") == "publisher":
                interval = int(cfg.get("publishIntervalMS", 1000))
                self.publish_intervals[instruction] = interval
                logger.debug(f"Instruction '{instruction}' set to publish every {interval} ms.")

    def _start_scheduled_publishers(self):
        for topic, interval in self.publish_intervals.items():
            if topic not in self.publish_threads:
                t = threading.Thread(
                    target=self._publish_topic_periodically,
                    args=(topic, interval),
                    name=f"Publisher-{topic}",
                    daemon=True
                )
                self.publish_threads[topic] = t
                t.start()
                logger.info(f"Started scheduled publisher for topic '{topic}' with interval {interval} ms.")

    def _publish_topic_periodically(self, topic: str, interval_ms: int):
        while not shutdown_flag.is_set():
            try:
                data = self._get_device_data(topic)
                self.latest_data[topic] = data
                self._publish_mqtt_data(topic, data)
                logger.debug(f"Periodic data published for '{topic}': {data}")
            except Exception as e:
                logger.error(f"Error in scheduled publishing for '{topic}': {e}")
            time.sleep(interval_ms / 1000.0)

    def _get_device_data(self, topic: str) -> Dict[str, Any]:
        # Simulate device data acquisition; replace with SDK/protocol logic
        t = int(time.time())
        if topic == "status":
            # Example: overall device status
            return {
                "timestamp": t,
                "device": self.device_name,
                "status": "OK",
                "channels": [{"id": i, "state": "active" if i % 2 == 0 else "idle"} for i in range(4)],
                "alarm": {"active": False, "code": 0},
                "config_health": "good"
            }
        elif topic == "decoder_channel_status":
            return {
                "timestamp": t,
                "channels": [{"id": i, "decode_state": "decoding"} for i in range(4)]
            }
        elif topic == "display_config":
            return {
                "timestamp": t,
                "layout": "4x4",
                "active_scene": "default"
            }
        else:
            # Generic
            return {
                "timestamp": t,
                "topic": topic,
                "data": "Sample data"
            }

    def _publish_mqtt_data(self, topic: str, data: Any, qos: int = 1):
        if not self.mqtt_client or not self.connected:
            logger.warning(f"MQTT client not connected; cannot publish topic '{topic}'")
            return
        mqtt_topic = f"{MQTT_TOPIC_PREFIX}/{self.device_name}/{topic}"
        try:
            payload = json.dumps(data, default=str)
            self.mqtt_client.publish(mqtt_topic, payload, qos=qos)
            logger.info(f"Published to MQTT topic '{mqtt_topic}': {payload}")
        except Exception as e:
            logger.error(f"Failed to publish to MQTT topic '{mqtt_topic}': {e}")

    def _run_http_server(self):
        app = self._flask_app

        @app.route("/health", methods=["GET"])
        def health():
            return jsonify({"status": "healthy"}), 200

        @app.route("/status", methods=["GET"])
        def status():
            try:
                status_data = self.latest_data.get("status", {})
                return jsonify(status_data), 200
            except Exception as e:
                logger.error(f"Failed to get status: {e}")
                return jsonify({"error": "Failed to get status"}), 500

        try:
            app.run(host=HTTP_HOST, port=HTTP_PORT, threaded=True)
        except Exception as e:
            logger.error(f"HTTP server failed: {e}")

    def _run_mqtt_loop(self):
        while not shutdown_flag.is_set():
            try:
                if not self.mqtt_client:
                    self.connect_mqtt()
                self.mqtt_client.loop_forever()
            except Exception as e:
                logger.error(f"MQTT loop error: {e}")
                time.sleep(5)

    def connect_mqtt(self):
        client = mqtt.Client(client_id=self.mqtt_client_id, clean_session=True)
        if MQTT_BROKER_USERNAME and MQTT_BROKER_PASSWORD:
            client.username_pw_set(MQTT_BROKER_USERNAME, MQTT_BROKER_PASSWORD)
        client.on_connect = self.on_mqtt_connect
        client.on_message = self.on_mqtt_message
        client.on_disconnect = self.on_mqtt_disconnect
        self.mqtt_client = client
        try:
            client.connect(MQTT_BROKER, MQTT_BROKER_PORT, keepalive=60)
            logger.info(f"Connecting to MQTT broker {MQTT_BROKER}:{MQTT_BROKER_PORT} as {self.mqtt_client_id}")
        except Exception as e:
            logger.error(f"MQTT connection failed: {e}")
            self.connected = False

    def on_mqtt_connect(self, client, userdata, flags, rc):
        if rc == 0:
            self.connected = True
            logger.info("MQTT connected successfully.")
            # Subscribe to control topics
            for control_topic in [
                f"{MQTT_TOPIC_PREFIX}/{self.device_name}/control/reboot",
                f"{MQTT_TOPIC_PREFIX}/{self.device_name}/control/playback"
            ]:
                try:
                    client.subscribe(control_topic, qos=1)
                    logger.info(f"Subscribed to control topic: {control_topic}")
                except Exception as e:
                    logger.error(f"Failed to subscribe to {control_topic}: {e}")
            # Subscribe to telemetry topic if needed
            telemetry_topic = f"{MQTT_TOPIC_PREFIX}/{self.device_name}/status"
            try:
                client.subscribe(telemetry_topic, qos=1)
                logger.info(f"Subscribed to telemetry topic: {telemetry_topic}")
            except Exception as e:
                logger.error(f"Failed to subscribe to {telemetry_topic}: {e}")
        else:
            logger.error(f"MQTT connect failed with rc={rc}")
            self.connected = False

    def on_mqtt_disconnect(self, client, userdata, rc):
        self.connected = False
        if rc != 0:
            logger.warning(f"Unexpected MQTT disconnection (rc={rc}). Attempting to reconnect.")
            try:
                client.reconnect()
            except Exception as e:
                logger.error(f"MQTT reconnect failed: {e}")
                time.sleep(5)

    def on_mqtt_message(self, client, userdata, msg):
        logger.info(f"MQTT message received: {msg.topic} {msg.payload}")
        try:
            payload = json.loads(msg.payload.decode('utf-8'))
        except Exception as e:
            logger.error(f"Failed to decode MQTT message: {e}")
            return

        if msg.topic.endswith("/control/reboot"):
            self._handle_reboot_command(payload)
        elif msg.topic.endswith("/control/playback"):
            self._handle_playback_command(payload)
        else:
            logger.warning(f"Unhandled topic: {msg.topic}")

    def _handle_reboot_command(self, payload: Dict[str, Any]):
        logger.info(f"Handling reboot command: {payload}")
        # Simulate reboot: update status, publish confirmation
        if payload.get("command") == "reboot":
            status = {"status": "rebooting", "timestamp": int(time.time())}
            self.latest_data["status"] = status
            self._publish_mqtt_data("status", status, qos=2)
            # Simulate reboot delay
            time.sleep(2)
            self.latest_data["status"] = {"status": "OK", "timestamp": int(time.time())}
            self._publish_mqtt_data("status", self.latest_data["status"], qos=1)
            logger.info("Device reboot simulated and status updated.")

    def _handle_playback_command(self, payload: Dict[str, Any]):
        logger.info(f"Handling playback command: {payload}")
        status = {"playback_action": payload.get("action"), "params": payload.get("params", {}), "timestamp": int(time.time())}
        # Simulate playback state update
        self.latest_data["playback"] = status
        self._publish_mqtt_data("playback", status, qos=1)

    def setup_routes(self):
        # Already set up in _run_http_server()
        pass

    def signal_handler(self, signum, frame):
        logger.info(f"Signal received: {signum}")
        shutdown_flag.set()
        self.shutdown()

    def shutdown(self):
        logger.info("Shutting down driver.")
        try:
            if self.mqtt_client:
                self.mqtt_client.disconnect()
                logger.info("MQTT client disconnected.")
        except Exception as e:
            logger.error(f"Error during MQTT disconnect: {e}")
        # Wait for threads to exit
        for t in self.publish_threads.values():
            t.join(timeout=2)
        if self._http_thread.is_alive():
            # Flask doesn't provide a native shutdown in threaded mode
            logger.info("HTTP server will terminate with main process.")
        logger.info("Shutdown complete.")

    def run(self):
        signal.signal(signal.SIGINT, self.signal_handler)
        signal.signal(signal.SIGTERM, self.signal_handler)
        self._http_thread.start()
        self._mqtt_thread.start()
        self._start_scheduled_publishers()
        logger.info("DeviceShifu MQTT Driver started.")
        try:
            while not shutdown_flag.is_set():
                time.sleep(1)
        except KeyboardInterrupt:
            logger.info("KeyboardInterrupt received, shutting down.")
            shutdown_flag.set()
        self.shutdown()

if __name__ == "__main__":
    driver = DeviceShifuMQTTDriver()
    driver.run()